felix/dataplane/linux/int_dataplane.go

```diff
@@ -1774,7 +1774,17 @@ func (d *InternalDataplane) setUpIptablesNormal() {
 			Match:  d.newMatch(),
 			Action: d.actions.Jump(rules.ChainNATPrerouting),
 		}})
-		t.InsertOrAppendRules("POSTROUTING", []generictables.Rule{{
+		// We must go last not to conflict with kube-proxy MASQ rules. All we do
+		// is MASQ anyway, but if kube-proxy set a mark to do MASQ, we must let it to
+		// clean it up so not to execute some rules after say a tunnel (vxlan)
+		// device - which should not happen.
+		//
+		// MASQ is a terminating target so once one is hit, no other rule gets
+		// executed.
+		//
+		// N.B. ChainFIPSnat does not do MASQ, but does not collide with k8s
+		// service, namely nodeports.
```

Comment:
```suggestion
		// We must go last to avoid a conflict if both kube-proxy and Calico 
		// decide to MASQ the traffic.  
		//
		// This is because kube-proxy uses a mark bit to trigger its MASQ and 
		// we need the mark bit to get cleared by kube-proxy's chain.  If we 
		// go first, our MASQ rule terminates further processing, and the 
		// mark bit remains set on the packet.  
		// 
		// Leaving the mark set on the packet is a problem when the packet 
		// gets encapped because the mark is copied to the outer encap packet.
		// The outer packet then gets MASQed by kube-proxy's rule.  In turn,
		// that MASQ triggers a checksum offload bug in the kernel resulting
		// in corrupted packets.
		//
		// N.B. ChainFIPSnat does not do MASQ, but does not collide with k8s
		// service, namely nodeports.
```